pkg/lockfile/lockfile_windows.go

```diff
@@ -5,89 +5,39 @@ package lockfile
 
 import (
 	"os"
-	"sync"
 	"time"
-)
-
-// createLockFileForPath returns a *LockFile object, possibly (depending on the platform)
-// working inter-process and associated with the specified path.
-//
-// This function will be called at most once for each path value within a single process.
-//
-// If ro, the lock is a read-write lock and the returned *LockFile should correspond to the
-// “lock for reading” (shared) operation; otherwise, the lock is either an exclusive lock,
-// or a read-write lock and *LockFile should correspond to the “lock for writing” (exclusive) operation.
-//
-// WARNING:
-// - The lock may or MAY NOT be inter-process.
-// - There may or MAY NOT be an actual object on the filesystem created for the specified path.
-// - Even if ro, the lock MAY be exclusive.
-func createLockFileForPath(path string, ro bool) (*LockFile, error) {
-	return &LockFile{locked: false}, nil
-}
 
-// *LockFile represents a file lock where the file is used to cache an
-// identifier of the last party that made changes to whatever's being protected
-// by the lock.
-//
-// It MUST NOT be created manually. Use GetLockFile or GetROLockFile instead.
-type LockFile struct {
-	mu     sync.Mutex
-	file   string
-	locked bool
-}
-
-// LastWrite is an opaque identifier of the last write to some *LockFile.
-// It can be used by users of a *LockFile to determine if the lock indicates changes
-// since the last check.
-// A default-initialized LastWrite never matches any last write, i.e. it always indicates changes.
-type LastWrite struct {
-	// Nothing: The Windows “implementation” does not actually track writes.
-}
-
-func (l *LockFile) Lock() {
-	l.mu.Lock()
-	l.locked = true
-}
-
-func (l *LockFile) RLock() {
-	l.mu.Lock()
-	l.locked = true
-}
-
-func (l *LockFile) Unlock() {
-	l.locked = false
-	l.mu.Unlock()
-}
-
-func (l *LockFile) AssertLocked() {
-	// DO NOT provide a variant that returns the value of l.locked.
-	//
-	// If the caller does not hold the lock, l.locked might nevertheless be true because another goroutine does hold it, and
-	// we can’t tell the difference.
-	//
-	// Hence, this “AssertLocked” method, which exists only for sanity checks.
-	if !l.locked {
-		panic("internal error: lock is not held by the expected owner")
-	}
-}
+	"golang.org/x/sys/windows"
+)
 
-func (l *LockFile) AssertLockedForWriting() {
-	// DO NOT provide a variant that returns the current lock state.
-	//
-	// The same caveats as for AssertLocked apply equally.
-	l.AssertLocked() // The current implementation does not distinguish between read and write locks.
-}
+const (
+	F_WRLCK   = windows.LOCKFILE_EXCLUSIVE_LOCK
+	F_RDLCK   = 0
+	O_CLOEXEC = windows.O_CLOEXEC
```

Comment:
(Aesthetically, I’m a bit unhappy with the generic code turning generic concepts like “read-write lock” into a UNIX-specific flag, only for the Windows implementation to pretend to be UNIX. It seems to me that the generic code could just pass a generic “read-only” flag and let the platform-specific handler select the right API.

But I didn’t actually check whether it would be easy to do it the other way… and the current approach is, I suppose, more readable for us Unix-native maintainers — and it is certainly less invasive. So I guess I don’t feel strongly about this.)

Reply:
@mtrmac I've push a commit which removes these constants in favour of a flag. Let me know if you feel this is a better approach

Reply:
Thanks!

Yes, I think that’s much cleaner.

Two more possible improvements:
- Use an enum for the lock type, instead of a `bool`. OTOH it’s used in few enough places that we can just pay attention.
- Use a platform-specific `fileHandle` type instead of universal `uintptr`; then the Windows code could use a `windows.Handle`.